### DISCUSSION TITLE
Is there ANY way to restore Folders & Tabs structure from the "Tree Tabs" extension?

### DISCUSSION THREAD

**Discussion by JohnVKRDam** [mbnuqw/sidebery#1196] [Q&A] [answered]
Hello,

I am currently using the extension "Tree Tabs" (this NOT the same as the "Tree Style Tab" extension!), but this extension is driving me absolutely INSANE. I lose my sessions, folder & tab structure on a VERY regular basis when I (have to) restart Firefox... and NONE of the options to restore the structure work... It's a TERRIBLE extension in that regard, to say the least, and I would like to move away from it in favor of Sidebery! I have been using/testing Sidebery (currently on v5.0.0rc4) in a Hyper-V machine for a while now and I have found it to be pretty stable?

I have searched for discussions on "import" and "tree tabs", for instance, but was unable to find anything relevant to my particular situation.

My question is:
Is there ANY way to convert the "Tree Tabs" folder structure to Sidebery?

I have many hundreds of Tabs open at any given time. Currently, for instance, about 1700, in several Groups. So, any manual operations for moving them to Sidebery are just impossibly hard.

I have tried using Bookmarks, because Sidebery can restore from Bookmarks. The Sidebery Bookmarks Restore option works perfectly well, but "Tree Tabs" just SUCKS (please excuse my language, it's really frustrating...) and always messes up the structure when saving to Bookmarks. So unfortunately, this is not an option I can use... :(

Is there any other way?

Link to the "Tree Tabs" extension: https://addons.mozilla.org/en-US/firefox/addon/tree-tabs

Thanks in advance!

Regards,
John

p.s. If everything else fails, is there a description of the structure of an import file for Sidebery? If so, I might be able to code my own solution in VS.NET. But to be able to do that, I would need to know how to write a proper file that allows import into Sidebery. If that is even possible?

**Comment by mbnuqw** [comment, marked as answer]
There is no simple way to import TT (Tree Tabs) data into sidebery and it will be probably faster to manually restucture it. But, you can try to form a Sidebery [snapshot](https://github.com/mbnuqw/sidebery/blob/358bcaf0218c624e22d4eac4d7aba9e6efd1cc86/src/types/snapshots.ts#L15) (in JSON) from TT data and import it (Sidebery settings / Snapshots viewer / Import snapshot). To get the TT data open `about:addons`, click on `gear` icon, click on `Debug Add-ons`, find Tree Tabs and click on `Inspect`, open multi-line console (ctrl+shift+K, ctrl+B).

```js
(async () => {
  const windows = await browser.windows.getAll({ populate: true })

  for (const win of windows) {
    const ttWinData = await browser.sessions.getWindowValue(win.id, 'TTdata')
    console.log(ttWinData)
    
    for (const tab of win.tabs) {
      const ttTabData = await browser.sessions.getTabValue(tab.id, 'TTdata')
      console.log(ttTabData)
    }
  }

  //... and then use this data to form the snapshot
})()
```

**Reply by JohnVKRDam** [reply]
That is helpful, thank you very much!

I have thousands of bookmarks (not all online, but also stored in "saved groups" from TT). Manually reconstructing them all would be months worth of nightmares ;) haha! So, thank you for the suggestion! I will definitely take a look at that.

Thanks again!
John

**Comment by gmevel** [comment]
In case it may still be of use to anyone: I have written a program that does exactly that: converting a TreeTabs dump (`*.tt_session`) to a Sidebery snapshot. It is [available here](https://gitlab.crans.org/mevel/scripts/-/blob/master/dump-treetabs-session). Note that it depends on file `json.ml` which is located in the same repo. To use the program, you need the OCaml language to be installed on your system; then:

    ./dump-treetabs-session --output=sidebery < MY_TREETABS_SESSION.tt_session > MY_SIDEBERY_SESSION.json
